sim/network.go

```diff
@@ -44,31 +49,77 @@ func newNetwork(opts *options) *Network {
 	}
 }
 
-func (n *Network) AddParticipant(p gpbft.Receiver) {
+func (n *Network) AddParticipant(p Receiver) {
 	if n.participants[p.ID()] != nil {
 		panic("duplicate participant ID")
 	}
 	n.participantIDs = append(n.participantIDs, p.ID())
 	n.participants[p.ID()] = p
 }
 
+////// Network interface
+
+func (n *Network) NetworkFor(signer gpbft.Signer, id gpbft.ActorID) *NetworkFor {
+	return &NetworkFor{
+		ParticipantID: id,
+		Signer:        signer,
+		Network:       n,
+	}
+}
+
+type NetworkFor struct {
+	ParticipantID gpbft.ActorID
+	Signer        gpbft.Signer
+	*Network
+}
+
+// Log implements tagging Tracer interface
+func (nf *NetworkFor) Log(format string, args ...interface{}) {
+	nf.Network.log(TraceLogic, "P%d "+format, append([]any{nf.ParticipantID}, args...)...)
+}
+
+func (nf *NetworkFor) RequestBroadcast(mt *gpbft.MessageTemplate) {
+	msg, err := mt.Build(nf.Signer, nf.ParticipantID)
+	if err != nil {
+		nf.Log("building message for: %d: %+v", nf.ParticipantID, err)
+		return
+	}
+	nf.log(TraceSent, "P%d ↗ %v", msg.Sender, msg)
+	for _, dest := range nf.participantIDs {
+		latencySample := time.Duration(0)
+		if dest != msg.Sender {
+			latencySample = nf.latency.Sample(nf.Time(), msg.Sender, dest)
+		}
+
+		nf.queue.Insert(
+			messageInFlight{
+				source:    msg.Sender,
+				dest:      dest,
+				payload:   *msg,
+				deliverAt: nf.clock.Add(latencySample),
+			})
+	}
+}
+
 func (n *Network) NetworkName() gpbft.NetworkName {
 	return n.networkName
 }
 
 func (n *Network) Broadcast(msg *gpbft.GMessage) {
 	n.log(TraceSent, "P%d ↗ %v", msg.Sender, msg)
 	for _, dest := range n.participantIDs {
-		if dest != msg.Sender {
-			latencySample := n.latency.Sample(n.Time(), msg.Sender, dest)
-			n.queue.Insert(
-				messageInFlight{
-					source:    msg.Sender,
-					dest:      dest,
-					payload:   *msg,
-					deliverAt: n.clock.Add(latencySample),
-				})
+		if dest == msg.Sender {
+			continue
```

Comment:
As far as I can tell this is functionally equivalent. If that's right, I think the previous approach without unstructured control flow is marginally better. Non-blocking.